SOLR-17582 Stream CLUSTERSTATUS API response

mlbiscoc opened the pull request:
https://issues.apache.org/jira/browse/SOLR-17582

# Description

`CLUSTERSTATUS` would aggregate and build the whole response including all collection properties into an `NamedList` before finally passing it to the response writers. With many or thousands of collections, this can use up memory.

# Solution

Instead of returning a `NamedList`, return a `MapWriter`. The Mapwriter will pull an `Iterator` from the documentStream and create `SolrCollectionProperiesIterator` implementing `iterator` and `override` `next()` to create and write the response in time, instead of writing the whole thing in memory.

# Tests

Fix appropriate tests that were using `NamedList` to `Map` because of the `MapWriter`.

# Checklist

Please review the following and check all that apply:

- [ ] I have reviewed the guidelines for [How to Contribute](https://github.com/apache/solr/blob/main/CONTRIBUTING.md) and my code conforms to the standards described there to the best of my ability.
- [ ] I have created a Jira issue and added the issue ID to my pull request title.
- [ ] I have given Solr maintainers [access](https://help.github.com/en/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork) to contribute to my PR branch. (optional but recommended, not available for branches on forks living under an organisation)
- [ ] I have developed this patch against the `main` branch.
- [ ] I have run `./gradlew check`.
- [ ] I have added tests for my changes.
- [ ] I have added documentation for the [Reference Guide](https://github.com/apache/solr/tree/main/solr/solr-ref-guide)

dsmiley commented on solr/core/src/java/org/apache/solr/handler/admin/ClusterStatus.java:
```diff
@@ -368,4 +341,77 @@ public static Map<String, Object> postProcessCollectionJSON(Map<String, Object>
     return collection;
   }
+
+  private class SolrCollectionProperiesIterator implements Iterator<NamedList<Object>> {
```
I was hoping we wouldn't need a custom Iterator.  We do need a method that takes in DocCollection (and some other context that is the same across collections) and returns a NamedList.  With such a method, we can call it via streamOfDocCollection.map(collState -> theMethod(collState, routeKey, liveNodes, etc.).iterator()

mlbiscoc replied:
Ah yeah don't know why I didn't do this first. Changed it appropriately.

dsmiley commented on solr/solrj/src/java/org/apache/solr/client/solrj/impl/BaseHttpClusterStateProvider.java:
```diff
@@ -140,11 +139,11 @@ private ClusterState fetchClusterState(SolrClient client)
     }
 
-    var collectionsNl = (NamedList<Map<String, Object>>) cluster.get("collections");
+    var collectionsNl = (Map<String, Map<String, Object>>) cluster.get("collections");
```
this var name is no longer appropriate.  Should be collectionsMap (and we assume by name).

dsmiley commented on solr/solrj/src/java/org/apache/solr/client/solrj/impl/BaseHttpClusterStateProvider.java:
```diff
@@ -185,7 +184,9 @@ private DocCollection fetchCollectionState(SolrClient client, String collection)
         submitClusterStateRequest(client, collection, ClusterStateRequestType.FETCH_COLLECTION);
 
-    var collStateMap = (Map<String, Object>) cluster.findRecursive("collections", collection);
+    var collState = (Map<String, Object>) cluster.findRecursive("collections");
```
so findRecursive can only resolve collections and not Map?  Seems worth a small enhancement to do so but I understand if you don't bother or defer.

mlbiscoc replied:
Made the small enhancement to NamedList and added a test to make sure it works. Lmk what you think.

dsmiley reviewed (commented):
Thanks for contributing!

mlbiscoc commented on solr/core/src/java/org/apache/solr/handler/admin/ClusterStatus.java:
```diff
@@ -368,4 +341,77 @@ public static Map<String, Object> postProcessCollectionJSON(Map<String, Object>
+      if (shard != null) {
+        String[] paramShards = shard.split(",");
+        requestedShards.addAll(Arrays.asList(paramShards));
+      }
```
Moved it out so its not per iteration.

mlbiscoc commented on solr/core/src/java/org/apache/solr/handler/admin/ClusterStatus.java:
```diff
@@ -368,4 +341,77 @@ public static Map<String, Object> postProcessCollectionJSON(Map<String, Object>
+
+      byte[] bytes = Utils.toJSON(clusterStateCollection);
+      @SuppressWarnings("unchecked")
+      Map<String, Object> docCollection = (Map<String, Object>) Utils.fromJSON(bytes);
```
So looked at this and I understand now why it was done this way. It wants to just write this out to the `TextWriter` as a normal POJO but that doesn't seem to be what the `DocCollection` class is. So instead what was done was to just write it to JSON `byte[]` then back to a `Map` which was the "easiest" way.

Was looking to avoid this back and forth, but there were a few options. I tried using an `ObjectMapper` to `Map.class` but there is an error for unable to cast Instant due to a missing dependency for Jackson we need to introduce.
`Java 8 date/time type `java.time.Instant` not supported by default Issue`

Other way is to introduce a some kind of `toMap` method so that the TextWriter can write this as a generic `Map`.

Another option which actually looks like the way we should go is I found that the `DocCollection` class extends ZkNodeProps which implements `MapWriter`. `DocCollection` already overrides `[writeMap](https://github.com/apache/solr/blob/9cef6e390719cbd7b55085cfef98fcb053785f77/solr/solrj/src/java/org/apache/solr/common/cloud/DocCollection.java#L427)` so we could just return this to the TextWriter! Unfortunately the ClusterStatus class does a bunch of JSON post processing such as `Health` added to the `Map` that the output is missing some things because of this `[postProcessCollectionJSON()](https://github.com/apache/solr/blob/9cef6e390719cbd7b55085cfef98fcb053785f77/solr/core/src/java/org/apache/solr/handler/admin/ClusterStatus.java#L334)` method. 

I am thinking we should refactor DocCollection to so we can just return this but the changes were much more drastic but may be worth it. Maybe in a different JIRA? This scope continues to creep with me adding improvement to `NamedList`. Would be happy to refactor this and pick it up if you agree. Would clean up much more code and avoid this JSON processing for every collection iteration.

mlbiscoc commented on solr/core/src/java/org/apache/solr/handler/admin/ClusterStatus.java:
```diff
@@ -205,54 +214,32 @@ private void fetchClusterStatusForCollOrAlias(
-            collectionStatus.put("aliases", collectionVsAliases.get(name));
+    MapWriter collectionPropsWriter =
+        ew -> {
+          Iterator<Map<String, Object>> it =
+              collectionStream
+                  .map(
+                      (collectionState) ->
+                          collectionPropsResponse(
+                              collectionState,
+                              collectionVsAliases,
+                              routeKey,
+                              liveNodes,
+                              requestedShards))
+                  .iterator();
+          while (it.hasNext()) {
+            Map<String, Object> props = it.next();
+            props.forEach(
+                (key, value) -> {
+                  try {
+                    ew.put(key, value);
+                  } catch (IOException e) {
+                    throw new RuntimeException(e);
+                  }
+                });
```
I am thinking, if we can change up DocCollection in another PR, we can delete all those post processing functions and condense the response to this.
```java
MapWriter collectionPropsWriter =
        ew -> {
          Iterator<DocCollection> it = collectionStream.iterator();
          while (it.hasNext()) {
            DocCollection collState = it.next();
            ew.put(collState.getName(), collState);
          }
        };
```

dsmiley replied:
I should mention Solr has two similar APIs, ClusterStatus & ColStatus.  See org.apache.solr.handler.admin.ColStatus#getColStatus.  It'd be awesome if there was a single method that takes a DocCollection (and some other params as needed) to produce a NamedList.  At least ColStatus's code doesn't have the sad JSON round-trip.  Maybe it's own PR or not as you wish.

mlbiscoc replied:
Gotcha. I'll take a look at this in a separate PR/Jira

dsmiley reviewed (commented):
Leave the round-trip serialization -- I didn't realize it was that way before.
Leave findRecursive be... not as easy as you thought and probably deserves its own issue.

dsmiley commented on solr/core/src/java/org/apache/solr/handler/admin/ClusterStatus.java:
```diff
@@ -205,54 +214,32 @@ private void fetchClusterStatusForCollOrAlias(
+                              routeKey,
+                              liveNodes,
+                              requestedShards))
+                  .iterator();
+          while (it.hasNext()) {
```
Oh; now that I see you are iterating it yourself (I had thought that would be driven externally), you shouldn't bother calling .iterator then looping; you can loop directly with stream.forEach

mlbiscoc commented on solr/core/src/java/org/apache/solr/handler/admin/ClusterStatus.java:
```diff
@@ -342,76 +356,47 @@ public static Map<String, Object> postProcessCollectionJSON(Map<String, Object>
+      DocRouter router = clusterStateCollection.getRouter();
+      Collection<Slice> slices = router.getSearchSlices(routeKey, null, clusterStateCollection);
+      for (Slice slice : slices) {
+        requestedShards.add(slice.getName());
```
This is a bug. I moved that old logic into here without realizing that parameter is manipulated. Changed `requestedShards` to `Set.of()` for immutability and made changes accordingly.

mlbiscoc commented on solr/core/src/java/org/apache/solr/handler/admin/ClusterStatus.java:
```diff
@@ -342,76 +356,47 @@ public static Map<String, Object> postProcessCollectionJSON(Map<String, Object>
+      PerReplicaStates prs = clusterStateCollection.getPerReplicaStates();
+      collectionStatus.put("PRS", prs);
     }
+    collectionProps.put(name, collectionStatus);
```
Yeah taking a deeper look, this extra map isn't needed. I kind of did it because of `crossCheckReplicaStateWithLiveNodes` but instead I modified that method and just return `collectionStatus`.

dsmiley commented on solr/core/src/java/org/apache/solr/handler/admin/ClusterStatus.java:
```diff
@@ -183,10 +181,12 @@ private void fetchClusterStatusForCollOrAlias(
+      requestedShards = Set.of(paramShards);
+    } else {
+      requestedShards = Set.of();
     }
```
I bet this whole thing could be one line using ternary operator and no line wrap. 
Also, if "shard" isn't specified, I think it's clearer to have requestedShards as null to indicate there was no request for any shards.

dsmiley commented on solr/core/src/java/org/apache/solr/handler/admin/ClusterStatus.java:
```diff
@@ -356,28 +346,28 @@ public static Map<String, Object> postProcessCollectionJSON(Map<String, Object>
       for (Slice slice : slices) {
-        requestedShards.add(slice.getName());
+        shards.add(slice.getName());
       }
```
something seems wrong to me here.  If requestedShards is specified, we should use that.  If routeKey is specified, we should use that (to compute the shards).  Or neither (99% of users won't do either).  But we shouldn't do both if for no other reason that it's confusing.

mlbiscoc replied:
We could throw a `SolrException(BAD_REQUEST)` higher up the stack if passed both `shard` and `_route_`. I can update the ref-docs to say you can only do one or the other. But this could break people who are doing both.

I also changed that logic into a single line and `forEach`

dsmiley replied:
I recommend that but understand if you leave out of scope.
(I really don't think this is going to break anyone's current usage, BTW)

mlbiscoc replied:
I'll add it in when I come back for the refactor.

dsmiley commented on solr/core/src/java/org/apache/solr/handler/admin/ClusterStatus.java:
```diff
@@ -295,22 +288,19 @@ private Map<String, Object> getCollectionStatus(
-              // node is not live, so this replica is actually down
-              replicaMap.put(ZkStateReader.STATE_PROP, Replica.State.DOWN.toString());
-            }
+    Map<String, Object> shards = (Map<String, Object>) collectionProps.get("shards");
+    for (Object nextShard : shards.values()) {
+      Map<String, Object> shardMap = (Map<String, Object>) nextShard;
+      Map<String, Object> replicas = (Map<String, Object>) shardMap.get("replicas");
+      for (Object nextReplica : replicas.values()) {
+        Map<String, Object> replicaMap = (Map<String, Object>) nextReplica;
```
I almost want to cry just glancing at this.
This is the poster-child for why Java introduced "var".  And there may be other approaches to improve it but that's the simplest.
(Yeah you didn't write it; I know)

dsmiley replied:
totally understood.  In some old Solr code like this, there's always "and one more thing" we could/should do but ultimately snowballs the scope out of control.  I leave it to you to do as you wish.  Thank you for your contribution here; I didn't mean to get more out of you than you bargained for :-)

dsmiley commented on solr/core/src/java/org/apache/solr/handler/admin/ClusterStatus.java:
```diff
@@ -216,28 +216,21 @@ private void fetchClusterStatusForCollOrAlias(
+                          liveNodes,
+                          requestedShards));
+                } catch (IOException e) {
+                  throw new RuntimeException(e);
```
Why does buildResponseForCollection throw IOException?  That's suspicious.

dsmiley replied:
putNoEx then

dsmiley commented on solr/core/src/java/org/apache/solr/handler/admin/ClusterStatus.java:
```diff
@@ -216,28 +216,21 @@ private void fetchClusterStatusForCollOrAlias(
-                  }
-                });
-          }
+          collectionStream.forEach(
+              (collectionState) -> {
```
friggin beautiful now

dsmiley reviewed (commented):
Uh oh; a back-compat problem is dawning on me.  Any SolrJ consumer of this API (especially using the default "javabin" format) is going to break here.  The change to BaseHttpClusterStateProvider should have made this glaringly obvious but I overlooked its implication earlier (face-palm).  

I wouldn't mind if we do this only for V2 (which doesn't yet have back-compatibility concerns) but V2 only has one collection status list returning API (/cluster) and I've debated against the very existence of /cluster for V2.  There's not yet a proposal for an alternative.  Furthermore V2 raises the bigger question of coalescing on a single DocCollection serialization method.

If we forge ahead anyway (not waiting for V2), we'll have to work around this compatibility.  I had a couple ideas but whittled down to one:
* detect the response format in this handler.  If it's Javabin, we write to a NamedList (or better, SimpleOrderedMap), defeating the point of this PR _for that format_.  If JSON/XML is used, this PR will accomplish its goal. Need to write a little test for JSON that at least partially sanity checks the result.  Such a test should pass without this PR.

mlbiscoc commented:
> Uh oh; a back-compat problem is dawning on me. Any SolrJ consumer of this API (especially using the default "javabin" format) is going to break here. The change to BaseHttpClusterStateProvider should have made this glaringly obvious but I overlooked its implication earlier (face-palm).
> 
> I wouldn't mind if we do this only for V2 (which doesn't yet have back-compatibility concerns) but V2 only has one collection status list returning API (/cluster) and I've debated against the very existence of /cluster for V2. There's not yet a proposal for an alternative. Furthermore V2 raises the bigger question of coalescing on a single DocCollection serialization method.
> 
> If we forge ahead anyway (not waiting for V2), we'll have to work around this compatibility. I had a couple ideas but whittled down to one:
> 
> * detect the response format in this handler.  If it's Javabin, we write to a NamedList (or better, SimpleOrderedMap), defeating the point of this PR _for that format_.  If JSON/XML is used, this PR will accomplish its goal. Need to write a little test for JSON that at least partially sanity checks the result.  Such a test should pass without this PR.

Well thats unfortunate... I'm not too familiar with SolrJ (I'll maybe play with it more to catch stuff like this) but I am assuming now the API is going to return a `Map` and no longer a `NamedList` to the client thus breaking them now?

I'll continue with this and add in your ideas to keep this backwards compatible.

dsmiley commented:
Yes; as the code you modified in a ClusterStateProvider shows, the NamedList->Map happened and broke that code.

You could add a check for {{params.get(CommonParams.WT)}} and check if "javabin" is the value.

mlbiscoc commented:
> Yes; as the code you modified in a ClusterStateProvider shows, the NamedList->Map happened and broke that code.
> 
> You could add a check for {{params.get(CommonParams.WT)}} and check if "javabin" is the value.

Added in that check for `javabin` to keep backwards compatibility with SolrJ and reverted all those SolrJ changes. Also created `clusterStatusWithCollectionAndShardJSON` to sanity check the `json` response from `MapWriter`. Happy holidays!
